link.go

```diff
@@ -0,0 +1,99 @@
+	Prototype() LinkPrototype
+
+	// String should return a reasonably human-readable debug-friendly representation the Link.
+	// There is no contract that requires that the string be able to be parsed back into a Link value,
+	// but the string should be unique (e.g. not elide any parts of the hash).
+	String() string
+}
+
+// LinkPrototype encapsulates any implementation details and parameters
+// necessary for creating a Link, expect for the hash result itself.
+//
+// LinkPrototype, like Link, is an interface in go-ipld-prime,
+// but the most common instantiation of it comes from the `linking/cid` package,
+// and represents CIDs (see https://github.com/multiformats/cid).
+// If using CIDs as an implementation, LinkPrototype will encapsulate information
+// like multihashType, multicodecType, and cidVersion, for example.
+// (LinkPrototype is analogous to cid.Prefix.)
+type LinkPrototype interface {
+	// BuildLink should return a new Link value based on the given hashsum.
+	// The hashsum argument should typically be a value returned from a
+	// https://golang.org/pkg/hash/#Hash.Sum call.
+	//
+	// The hashsum reference must not be retained (the caller is free to reuse it).
+	BuildLink(hashsum []byte) Link
+}
+
+// LinkContext is a structure carrying ancilary information that may be used
+// while loading or storing data -- see its usage in BlockReadOpener, BlockWriteOpener,
+// and in the methods on LinkSystem which handle loading and storing data.
+//
+// A zero value for LinkContext is generally acceptable in any functions that use it.
+// In this case, any operations that need a Context will use Context.Background
+// (thus being uncancellable) and simply have no additional information to work with.
+type LinkContext struct {
+	// Ctx is the familiar golang Context pattern.
+	// Use this for cancellation, or attaching additional info
+	// (for example, perhaps to pass auth tokens through to the storage functions).
+	Ctx context.Context
+
+	// Path where the link was encountered.  May be zero.
+	//
+	// Functions in the traversal package will set this automatically.
+	LinkPath Path
+
+	// When traversing data or encoding: the Node containing the link --
+	// it may have additional type info, etc, that can be accessed.
+	// When building / decoding: not present.
+	//
+	// Functions in the traversal package will set this automatically.
+	LinkNode Node
+
+	// When building data or decoding: the NodeAssembler that will be receiving the link --
+	// it may have additional type info, etc, that can be accessed.
+	// When traversing / encoding: not present.
+	//
+	// Functions in the traversal package will set this automatically.
+	LinkNodeAssembler NodeAssembler
+
+	// Parent of the LinkNode.  May be zero.
+	//
+	// Functions in the traversal package will set this automatically.
+	ParentNode Node
+
+	// REVIEW: ParentNode in LinkContext -- so far, this has only ever been hypothetically useful.  Keep or drop?
```

Comment:
I've ended up squashing most of my work-in-progress commits together, so it's now invisible, but it might be interesting to note: I originally did comment this out entirely... but by the time I finished migrating the `traversal` package, it turned out so easy to keep it that I did.

So, this doesn't really argue against dropping it.  But it turned out not much work (at the moment, anyway) to keep it, either.